src/Cardano/Wallet/CoinSelection.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RankNTypes #-}
+
+{-# OPTIONS_GHC -fno-warn-redundant-constraints #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Provides the API of Coin Selection algorithm and Fee related functionality.
+-- For more information refer to:
+-- https://iohk.io/blog/self-organisation-in-coin-selection/
+
+
+module Cardano.Wallet.CoinSelection where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types
+    ( Coin, Tx, TxIn, TxOut, UTxO )
+import Control.Monad.Random
+    ( MonadRandom )
+import Control.Monad.Trans.Except
+    ( ExceptT (..) )
+import Data.ByteString
+    ( ByteString )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+
+data CoinSelectionOptions = CoinSelectionOptions
+    { estimateFee
+        :: Int
+        -> NonEmpty Coin
+        -> Coin
+        -- ^ A function to estimate the fees.
+    , dustThreshold
+        :: Coin
+        -- ^ Change addresses below the given threshold will be evicted
+        -- from the created transaction. If you only want to remove change
+        -- outputs equal to 0, set 'csoDustThreshold' to 0.
+    , maximumNumberOfInputs
+        :: Word64
+    } deriving (Generic)
+
+
+data ErrCoinSelection = ErrCoinSelection
+
+data CoinSelFinalResult = CoinSelFinalResult
+    { inputs  :: NonEmpty (TxIn, TxOut)
+    -- ^ Picked inputs
+    , outputs :: NonEmpty TxOut
+    -- ^ Picked outputs
+    , change  :: [Coin]
+    -- ^ Resulting changes
+    } deriving (Show, Generic)
+
+
+-- | Random input selection policy
+random
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> UTxO
+    -> ExceptT ErrCoinSelection m CoinSelFinalResult
+random _ = undefined
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> UTxO
+    -> ExceptT ErrCoinSelection m CoinSelFinalResult
+largestFirst _ = undefined
+
+
+----------------------------------------------------------------------------
+--                        Fee related                                     --
+----------------------------------------------------------------------------
+
+newtype Fee = Fee { getFee :: Quantity "lovelace" Natural }
+
+estimateCardanoFee :: Tx -> Word64
+estimateCardanoFee _ = 1765
+
+estimateMaxTxInputs
+  :: ByteString
+    -- ^ Maximum size of a transaction
```

Comment:
size of tx in bytestring - how should one interpret this ?

Reply:
you are right, should be `Bytes` (memory unit), like in http://hackage.haskell.org/package/serokell-util-0.10.0/docs/src/Serokell.Data.Memory.Units.html#Byte

Reply:
I wonder if we are ok with using `serokell-utils` or better define `Bytes` ourselves?

Reply:
It's okay to use libraries (from Serokell or anyone) when they are sound and solve an actual problem. Here, `serokell-util` isn't really sound nor useful for most of the types or functions it provides IMO. 

Here, a size could just be a `Quantity "byte" Word32` if you really want to disambiguate the type. Yet, I believe it's not needed to implement largest first :thinking: ?

Reply:
if it is not needed in the end I will definitely remove it. Otherwise, will rely on `Quantity "byte" Word32`